### ISSUE TITLE
Literal key check in a TypedDict with optional keys

### ISSUE THREAD

**Issue by AndreasMadsen** [microsoft/pyright#9727]
**Describe the bug**

After checking a `key` is in a `TypedDict` with `total=False` and the key is a `Literal`, then `dict[key]` reports incorrectly that the key is not guaranteed. A similar issue was reported in https://github.com/microsoft/pyright/issues/7589 and was correctly fixed. The difference is that it was fixed for `str` and not for `Literal[str]`. If the `key` is not annotated with `Literal` but instead with `str` the example does not display the bug.

**Code or Screenshots**

```python
from typing import TypedDict, Literal

class Test(TypedDict, total=False):
    a: int
    b: int

a: Test = { 'a': 1 }

def get(key: Literal['a', 'b'], dict: Test) -> int:
    if key not in dict:
        raise ValueError(f'missing key {key}')
    return dict[key] # reportTypedDictNotRequiredAccess
```

```
Could not access item in TypedDict
  "a" is not a required key in "Test", so access may result in runtime exception
  "b" is not a required key in "Test", so access may result in runtime exception
Pylance[reportTypedDictNotRequiredAccess](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportTypedDictNotRequiredAccess)
```

**VS Code extension or command-line**

VSCode with Pylance version 2024.12.1 and 2025.1.100 (pre-release)

**Comment by erictraut** [comment]
Pyright is working as intended here, so this isn't a bug. I'll change it to a feature request and consider it in that light.

Pyright also support the narrowing of `D` in type guard pattern `S in D` where `D` is a TypedDict and `S` is a string literal. This pattern doesn't apply when the first argument is a variable. If you change your code to the following, pyright's type narrowing will work as you expect.

```python
    if "a" not in dict or "b" not in dict:
        raise ValueError(f"missing key {key}")
```

For more details about what is and is not supported for type narrowing patterns, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards).

It would be theoretically possible to extend pyright's type narrowing logic to accommodate the pattern you're using here, but it's unlikely we'd do so unless we received a strong signal from other pyright users that this is a common pattern. Each new type guard pattern requires custom logic and has the potential to slow down overall type analysis, so the bar is relatively high. So far, you're the only one who has requested this. Plus, mypy (another static type checker) doesn't support any form of type narrowing for TypedDict keys, so pyright already goes well beyond other type checkers in this area.

I'm going to reject the feature request for now, but I'm willing to revisit this in the future if there is sufficient interest (in the form of upvotes to this issue).

**Comment by Daraan** [comment]
+1 for this feature.

I claim its a common pattern to perform a "key in my_dict" (S in D) if the key might not be present. That after such a check a `reportTypedDictNotRequiredAccess` is encountered is weird to me, as I verified that the key is present.

The best solution currently to me is to combine the access (after the key check) with a `# pyright: ignore[reportTypedDictNotRequiredAccess]`

Code sample in [pyright playground](https://pyright-play.net/?pythonVersion=3.13&strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgCYAiSAxjADRRUCGAznVCMQG7GMA2A%2BvATEAsAChxVXi2aFyACiLCK1NjDAw%2BAXgBifZsQCUALnFRzURlGOYUMMxYBG1qKxDjxAYigBVZqgwmVigAC2IOegB3JBgQ2FIbAnIoLSgAbwAiRgybAEYAXygEOBAkNBD8AGsUMEiUWVjGfGckWRr8BA4DO3EYUhSGFhhFcnpM7Lz8ww8xYFwoSuI4Wyh5LIz6DMcMkwdzJGAFpZW%2B0zELC-YuHgEhYnk%2BgG1FuABdQ3NvDhw8JTJKGgAOQ0ACViABHACuSA4pAAglQqMRmMw9lA0d4APKVeiOSH4dqWVxgXj4pBgFAAOjRByOyy0qXWlhQ-ReJwSaIunVQwyeL3enygYEqAG4oHj8JCDLxkbJSMiyigik0YOEUGiMVAMgBlElkikZKC5egyRWYfDqImcPiQ4jM-qMRHEBD4KyK3Cic4WWls1DxM6XczW0l21J8pavQXFUrlGA2d0cR7fXAwP4qIGgiHQ2EIpEo16c8wcbh8QQkeTB23TCRewU6vXqFCGgBMUCldqsaGIbAlUF4YAMsBItjmIAgTXJ6trUErofilK7wxe9GBKCMNMOs8wslXxADgYYFJgqFtaOLNzLwgrNqMgrhKGWxgA9DMLt566TjwaoABmehBKpjktKwaiVXgYnCPgxV7ABJPsBztO4R1wccvynC42VSAD5DcegXmrC4fWOP1TkLItrlLO4HlIZ4Iw%2BKBvB8FBqlqJVnxmLEAGlxXxVwYBARh0DteV-DQJUEBVNV7SgQlgB4fxHBlKBRygccHzpVExFIMB%2BFYSFgEOVJz0okhxFpJkSI5adtN0mB9OAajHnWd5xGIcDDi2Q1LP3cwbL0gzHM8lyvSAA)

```python
from typing import TypedDict, cast, reveal_type

class TD(TypedDict, total=False):
    a : int
    b : str

# Using cast here, with td: TD = {"a": 1} pyright knowns that b is not present
td = cast(TD, {"a": 1})

for key in ("a", "b"):
    if key in td:
        reveal_type(td[key])  # reportTypedDictNotRequiredAccess
    
    # Ok, but not a solution.
    if key == "a" and key in td:
        print(td[key])  # ok; but useless design pattern
    
    # "Solution" 1, assign it to a value and accept a ignore
    if key in td:
       value = td[key]  # pyright: ignore[reportTypedDictNotRequiredAccess]
       reveal_type(value)

    # "Soluton" 2 use a get, but lose type information
    value = td.get(key, None)
    if value is None:
        continue
    reveal_type(value)  # Any :/

    # "Solution" 3, cast key to a non literal; but I lose type information
    key = cast(str, key)
    if key in td:
        reveal_type(td[key])  # Unknown :/

# OK but strange design pattern and not feasible for many keys
do_stuff = reveal_type
if "a" in td:
    do_stuff(td["a"])
elif "b" in td:
    do_stuff(td["b"])
  
```

**Comment by AndreasMadsen** [comment]
I have thought about this, and I'm not really convinced "feature request" is appropriate. The code pattern is common (`variable in dict`) and the key is guaranteed to exist, yet an error is produced. I could understand if the key lookup just assumed the key existed, as if the dict type was `dict[str, int]`. However, an error for correct code is a bug.

For the record, I believe this comment also raised this issue: https://github.com/microsoft/pyright/issues/9338#issuecomment-2443627801
